Add readme instructions, repos and CMake toolchain file

lmayencourt opened the pull request.

lmayencourt committed:
Add repos file and CMake toolchain file

ROS2 "release-latest" branch is used.

lmayencourt committed:
Add readme instructions

Signed-off-by: Louis Mayencourt <louis.mayencourt@arm.com>

mikaelarguedas commented:
As these files are copied from another place it would be better to maintain the history of these files.

The repos file is using specific tags and hashes pointing to the Ardent release. Note that this means that people using this repos file will not be able to get new features or bug fixes by running `vcs pull` (but maybe pinning down specific versions is done on purpose here, I just wanted to confirm).

davwan01 commented:
@mikaelarguedas , thanks for your review. :-)
Are you suggesting to clone ros2/ros2 and update on top of that?

lmayencourt commented:
@mikaelarguedas thanks for your inputs, 
I create the `repos` file based on the `release-latest` branch, but it is probably better to point to `master` !

mikaelarguedas commented:
> Are you suggesting to clone ros2/ros2 and update on top of that?

Yeah there are several ways to do this:
- fork ros2/ros2 here and to build on top.
- push all the ros2/ros2 history without forking and build on top
- use git-filter-branch (example [here]( https://blogs.s-osg.org/retain-history-moving-files-git-repositories/)) to keep the history only of the files you want to keep.

I think it really depends what the end goal is here.
For the repos file: Is the purpose of this repos file to later be modified to add new repositories hosted here ? to remove some ?
If the goal is to add new repositories, I don't think there is a need to copy the repos file from ros2/ros2 (as it means more effort to maintain as each time something changes on the ros2/ros2 version we would need to change it here as well). We could have a repos file here only for the additional repositories, and the instructions can just be:
- `wget the ros2/ros2.repos`
- `wget the ros2-for-arm/ros2-for-arm.repos`
- `vcs import src < ros2.respo ros2-for-arm.repos`

For the toolchain file:
My guess is that it will be hosted on this repository and removed from the ros2/ros2 right ? in that case I'll just push it here with its' complete history and remove it from ros2/ros2 (and add a note in the instructions saying "if you want to cross compile go to the readme of ros2-for-arm/ros2 (link)"

davwan01 commented:
@mikaelarguedas , Thanks for sharing the detailed approaches. 

The purpose of this repository: I think the answer is "both".

- Add new repositories: we will have some new repositories which are for Arm platform or Arm specific features, e.g. secure message demo, DDS security API/library. For these, we can make this repo file as the second parameter for vcs-import as you suggested. If something is common, we can make PR to ros2/ros2 also.
- Remove some repositories: Our work will be based on "cross-compile(cc)" because honestly, it's more common to use x86 not Arm desktop/server to develop ros2 applications at the moment. We have to remove some packages to make cc working for now. In that case, we kept the whole manifest here. But our target is to make cc working with all packages listed in ros2/ros2.  **So my proposal is to give instructions (in README) on how to get and modify the ros2/ros2 manifest file and we just keep the additional packages and toolchain file here.**

For toolchain file:
My understanding is that "ros2/ros2" is the common manifest file which is platform independent. "ros2-for-arm/ros2-for-arm" will be the additional manifest to facilitate the development of Arm platforms.
So I think you're right, we can move the toolchain from ros2/ros2 to here.
**In that case, it seems we don't need to fork ros2/ros2.**

davwan01 commented on ros2-arm.repos:
```diff
@@ -0,0 +1,209 @@
+repositories:
```
It seems ros2-for-arm.repos is a better name?

lmayencourt replied:
My bad, I forgot to update the name before pushing !
`ros2.repos` or `ros2-for-arm.repos` ? Since the file is already under the ros2-for-arm organisation.

sudpau01 commented:
@lmayencourt 
1.readme.md in this git repository will just point to the wiki page (https://github.com/ros2-for-arm/ros2/wiki/ROS2-on-arm-architecture)
This wiki will essentially have the contents of https://github.com/ros2/ros2/blob/master/arm_crosscompilation/README.md 
Finally remove the  https://github.com/ros2/ros2/blob/master/arm_crosscompilation/README.md 

2.Move https://github.com/ros2/ros2/blob/master/arm_crosscompilation/aarch64_toolchainfile.cmake as a part of this git repository. (which is already done in this pull request). It could be removed from ros2/ros2 

3.rename to ros2-for-arm.repos

@mikaelarguedas , @davwan01 agree?

sudpau01 commented:
https://github.com/ros2-for-arm/ros2/wiki/ROS2-on-arm-architecture

has being put. Please review

lmayencourt committed:
Update according to PR comments

ros2-arm.repos:
- rename ros2-arm.repos to ros2-for-arm.repos
- include only the ros2-for-arm repository

README.md
- move the instructions to wiki
- point to wiki

Signed-off-by: Louis Mayencourt <louis.mayencourt@arm.com>

mikaelarguedas commented:
Looks better now, thanks for iterating!

Note that the new wiki page points to "release-latest" for the ros2.repos file (which is totally fine if you want to target a stable version rather than master).

Can you provide more details about why the gnu toolchain is not sufficient and a specific version of the linaro toolchain is required instead. That's how it was before and as this is a hard-coded link, [the instructions broke](https://github.com/ros2/ros2/pull/439) as soon as Linaro changed the URL, so it doesnt seem very future-proof, especially given that current and upcoming ubuntu will have a more recent toolchain and compiler version than the one linked in this PR (see https://packages.ubuntu.com/bionic/g++-6-aarch64-linux-gnu).

Can you please open a matching PR on ros2/ros2 to remove the files in the "arm_crosscompilation" folder and just leave a README pointing to the wiki page of this repository ?

lmayencourt committed:
Change the link of the wiki to the 'ROS2-on-arm-architecture' page

Signed-off-by: Louis Mayencourt <louis.mayencourt@arm.com>

sudpau01 commented:
@mikaelarguedas 

We will like to stay pointing to the latest release version as we know it will work with the cross-compilation for ARM architecture. Ideally we could be pointing to master but we never know any package addition into the master ros2.repo might break the cross compilation. This is essentially because we don't have a CI setup where we could test regularly the cross compilation of the stack.

gnu toolchain should be sufficient.But we have the toolchain from Linaro that is customized and optimized one, you could consider it as the latest toolchain for ARM. But we will test it on our side and then if the gnu toolchain is sufficient then we will point to that in our wiki...

I think Luis already sent a PR for removing the cross-compile.cmake file and also updated the readme.md to point to our latest wiki page

mikaelarguedas commented:
> We will like to stay pointing to the latest release version as we know it will work with the cross-compilation for ARM architecture. Ideally we could be pointing to master but we never know any package addition into the master ros2.repo might break the cross compilation. This is essentially because we don't have a CI setup where we could test regularly the cross compilation of the stack.

Sounds good to me :+1: 

> I think @lmayencourt already sent a PR for removing the cross-compile.cmake file and also updated the readme.md to point to our latest wiki page

Yes that's be done [here](https://github.com/ros2/ros2/pull/466) as a follow-up of my previous comment :+1: 

> gnu toolchain should be sufficient.But we have the toolchain from Linaro that is customized and optimized one, you could consider it as the latest toolchain for ARM. But we will test it on our side and then if the gnu toolchain is sufficient then we will point to that in our wiki...

Understood, yeah it would be great to test that the gnu one works as well or at least to check on a regular basis that the link is still valid to make sure to have working instructions for users

mikaelarguedas reviewed (approved):
lgtm, thanks for adressing the comments

lmayencourt committed:
Update the name of the arm example repo

davwan01 commented on aarch64_toolchainfile.cmake:
```diff
@@ -0,0 +1,31 @@
+# Copyright (c) 2017, ARM Limited. All rights reserved.
```
Please use SPDX format:
# Copyright (c) 2018, ARM Limited.
# SPDX-License-Identifier: Apache-2.0

lmayencourt committed:
Change the license to SPDX format
